### ISSUE TITLE
XCP-ng PV-Drivers 8.2.2

### ISSUE THREAD

**Issue by borzel** [xcp-ng/xcp#181]

**Comment by borzel** [comment]
aahhh!!!! could not build xeniface because of a bug in Windows SDK... ahhh!!!!  maybe now it builds ...

**Comment by borzel** [comment]
build of PV Drivers 8.2.2 was successfull
Now I test on a german Windows 10 Pro x64 to install all the fun stuff, if this goes well, we have the next beta release :-)

**Comment by borzel** [comment]
for our brave tester: https://github.com/xcp-ng/win-pv-drivers/releases/tag/v8.2.2-beta

**Comment by stormi** [comment]
Thanks @borzel :+1:

**Comment by borzel** [comment]
Found the reason why the drivers get not installed when the main installer runs:

![grafik](https://user-images.githubusercontent.com/5782951/60385196-d3219780-9a86-11e9-87de-3f45486342a5.png)

Searching how to resolve this...

**Comment by borzel** [comment]
So if you install the drivers later (like mentioned in some troubleshooting guides and posts) this pops up:

![grafik](https://user-images.githubusercontent.com/5782951/60385317-2ea05500-9a88-11e9-8ec2-167f3c72e424.png)

and this is in the logfile `C:\Windows\INF\setupapi.dev.log`

![grafik](https://user-images.githubusercontent.com/5782951/60385362-ad958d80-9a88-11e9-98f2-3a5848bb1278.png)

... :thinking: ...

**Comment by borzel** [comment]
Hmm... after some reading .. all the relevant things are there:
* valid Certificate Chain - from Certum
* valid Timestamp - from Symantec :-/
* valid Cross Certificate Chain - from Microsoft

![grafik](https://user-images.githubusercontent.com/5782951/60386084-24cf1f80-9a91-11e9-8cbe-8b4329600ff6.png)

**Comment by borzel** [comment]
Here someone has the same problem, but with Windows 7 (*still reading*): 

https://social.msdn.microsoft.com/Forums/office/en-US/b65222ea-d11a-41cd-8d9b-fef43aedb065/howto-sign-kernel-driver-for-automated-field-deployment?forum=wdk

**Comment by borzel** [comment]
Our `.cat` file maybe is not properly signed:

![grafik](https://user-images.githubusercontent.com/5782951/60386406-aeccb780-9a94-11e9-831c-59976cdde6a8.png)

**Comment by borzel** [comment]
ahhh!!!! :bulb: 

If you manually acknowledge the driver signature and the checkbos (trust sfotware from "Vates") it installs the public cert into your personal certificate store.

As I worked on the installer I wondered why CITRIX installed his public cert in the certificate store ... but maybe it's for that exact reason :man_facepalming: so that the driver installing can work silently in the main installer.

Oh man (and all woman)! MS and documenting the relevant stuff is hard...

**Comment by borzel** [comment]
Installed the Vates cert in the trusted publisher cert store:

![grafik](https://user-images.githubusercontent.com/5782951/60386774-ea1db500-9a99-11e9-834d-5f1982628a69.png)

another problem occures:

![grafik](https://user-images.githubusercontent.com/5782951/60386760-bc387080-9a99-11e9-9a3d-5d7bc407907e.png)

So maybe the `.cat` file is not created properly.

**Comment by borzel** [comment]
Ahh!!! :bulb: 

The cert has to be in the computers trusted publishers store!

![grafik](https://user-images.githubusercontent.com/5782951/60386822-a2e3f400-9a9a-11e9-9ccf-d3fd33a75141.png)

**Comment by borzel** [comment]
There is the relevant function (https://github.com/xcp-ng/win-installer/blob/d232c8e5b074f1137f83b1d9d0d858fc727ea790/src/HelperFunctions/Helpers.cs#L88): 

![grafik](https://user-images.githubusercontent.com/5782951/60387140-57801480-9a9f-11e9-86d9-6869a3985941.png)

And there our cert is/has to be mentioned (https://github.com/xcp-ng/win-installer/blob/d232c8e5b074f1137f83b1d9d0d858fc727ea790/src/InstallAgent/InstallAgent.cs#L361):

![grafik](https://user-images.githubusercontent.com/5782951/60387194-dd9c5b00-9a9f-11e9-9bf4-d411bc766d02.png)

**Comment by borzel** [comment]
Release Candidate is here: https://github.com/xcp-ng/win-pv-drivers/releases/tag/v8.2.2.200-RC1

Please test and report back!

**Comment by sanmigit** [comment]
Management Agent fails to install on Windows Server Standard 2019 Spanish (Spain) (probably cosmetic)
Event viewer complains about unmanaged system exception:

Aplicación: InstallAgent.exe
Versión de Framework: v4.0.30319
Descripción: el proceso terminó debido a una excepción no controlada.
Información de la excepción: System.Exception
   en HelperFunctions.Helpers.BitIdxFromFlag(UInt32)
   en PVDevice.PVDevice.AllFunctioning()
   en InstallAgent.InstallAgent.__InstallThreadHandler()
   en InstallAgent.InstallAgent.InstallThreadHandler()
   en System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   en System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   en System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   en System.Threading.ThreadHelper.ThreadStart()

![image](https://user-images.githubusercontent.com/36620837/60571185-63a3f480-9d73-11e9-99b6-954d7e5c5a23.png)

Despite the error, my Xenserver 7.2 reports:
I/O optimized
Management Agent installed
Not able to receive updates from Windows Update

And after running C:\Program Files\XCP-ng\XenTools\install-XenProvider.cmd my list of vss providers reports XCP-ng XEN VSS Provider flawlessly:

C:\Program Files\XCP-ng\XenTools>vssadmin list providers
vssadmin 1.1 - Herramienta administrativa de línea de comandos del Servicio de instantáneas de volumen.
(C) Copyright 2001-2013 Microsoft Corp.

Nombre de proveedor 'Microsoft File Share Shadow Copy provider'
   Tipo de proveedor: Recurso compartido de archivos
   Id. de proveedor: {89300202-3cec-4981-9171-19f59559e0f2}
   Versión: 1.0.0.1

Nombre de proveedor 'Microsoft Software Shadow Copy provider 1.0'
   Tipo de proveedor: Sistema
   Id. de proveedor: {b5946137-7b9f-4925-af80-51abd60b20d5}
   Versión: 1.0.0.7

Nombre de proveedor 'XCP-ng XEN VSS Provider'
   Tipo de proveedor: Hardware
   Id. de proveedor: {3aeb8223-a8eb-43a2-8ff7-868312e67a8f}
   Versión: 1.0

**Comment by borzel** [comment]
@sanmigit from what I see on your screenshot: you just have to reboot, this is a known error :-)

**Comment by nagilum99** [comment]
Is that related to the bug, which appears only if language is not english?

**Comment by borzel** [comment]
@nagilum99 no, that language bug is only related to Windows Update. We don't have XCP-ng packages in windows update (yet)

**Comment by imtrobin** [comment]
Testing XCP-ng.Windows.PV.Drivers.8.2.2.200-rc1 on XCP-NG 7.6.0, win10 LTSC 64bit, english.

It works great, thanks!

**Comment by imtrobin** [comment]
Sorry, I spoke too fast. I did get one instance installed fine, but I cannot repeat it in subsequent installs. 
In other VMs
1. First install 8.2.2.200-rc1, will result"Management Management Agent Failed to install", 
2. Install again with Repair, it will show Windows Management Agent installed successfully'
3. Reboot, but Windows Management Agent is not installed.

I still have the instance where the Windows Management Agent is working fine ,but I can't reproduce it today.

**Comment by olivierlambert** [comment]
Same Windows 10 64bits?

**Comment by imtrobin** [comment]
Yes, same win 10 LTSC. Actually I'm trying two different machines, fresh win10 VMs, same results.

Both will get BSOD StopCode: INACCESSIBLE BOOT DEVICE after I install XCP-ng.Windows.PV.Drivers.8.2.2.200-rc1

I think win10 seems to be installing a version itself , maybe from windows update? The strangest thing is the working LTSC from yesterday is working fine, but today new installs will BSOD.

**Comment by olivierlambert** [comment]
I'm downloading Win 10 LTSC right now for my lab and I'll report back

**Comment by olivierlambert** [comment]
So I did a fresh install of Win 10 US LTSC 64 bits.

Here is the device manager after the first boot:

![Capture d’écran de 2019-07-12 12-31-12](https://user-images.githubusercontent.com/1241401/61122170-5dafc100-a4a1-11e9-8e07-88e891492fe4.png)

As you can see, all emulated drivers. Then I downloaded the ZIP file or RC drivers, extracted and run "setup":

![Capture d’écran de 2019-07-12 12-31-44](https://user-images.githubusercontent.com/1241401/61122210-73bd8180-a4a1-11e9-9f28-969ead0912a3.png)

Then after reboot I had:

![Capture d’écran de 2019-07-12 12-32-35](https://user-images.githubusercontent.com/1241401/61122221-7cae5300-a4a1-11e9-99cc-b07f4318fbce.png)

And finally, the device manager now:

![Capture d’écran de 2019-07-12 12-32-52](https://user-images.githubusercontent.com/1241401/61122235-82a43400-a4a1-11e9-9de0-5ea50c513d22.png)

Also the management agent is running well:

![Capture d’écran de 2019-07-12 12-35-18](https://user-images.githubusercontent.com/1241401/61122255-9059b980-a4a1-11e9-8524-cad4f5a1b9db.png)

![Capture d’écran de 2019-07-12 12-35-28](https://user-images.githubusercontent.com/1241401/61122264-951e6d80-a4a1-11e9-88df-60bed202787d.png)

**Comment by olivierlambert** [comment]
Also some disk bench showing PV drivers are here:

![Capture d’écran de 2019-07-12 12-48-20](https://user-images.githubusercontent.com/1241401/61122965-6bfedc80-a4a3-11e9-89a7-814f4266fcaa.png)

**Comment by imtrobin** [comment]
Repro Test 1
1. Fresh Install. The device manager looks the same as yours.
2. Reboot. I get the popup XenServer PV Storage Host needs to restart. I did not install the guest tools yet.
![XCP-ng Center 7 6 1-2019-07-13 11_36_12](https://user-images.githubusercontent.com/531606/61166587-aaa79c00-a562-11e9-857b-ee25aa6a9508.jpg)
3.Reboot. The device manager already shows XenSRC PVDisk
4. Now I also get the popup, Xenserver PV Network Class needs to restart. I did not yet install guest tools.
![image](https://user-images.githubusercontent.com/531606/61166605-dd519480-a562-11e9-9996-00c788d349b8.png)
5. Reboot. It already has the this installed without me installing the guest tools
![image](https://user-images.githubusercontent.com/531606/61166637-57821900-a563-11e9-8ba8-8ced6db86885.png)
6. Install guest tools. Fail to install guest management tools.
7. Reboot. BSOD
![image](https://user-images.githubusercontent.com/531606/61166662-bf386400-a563-11e9-8173-5265415098af.png)



Repro Test 2
1 Fresh Install.
2. Reboot. I get the popup XenServer PV Storage Host needs to restart. I press Cancel and install tools
![image](https://user-images.githubusercontent.com/531606/61166631-34f00000-a563-11e9-8eea-eaeaa097c52e.png)
3. Run setup again. Still fail to install guest managemant tools
4.Reboot. BSOD


The one that worked, I might have installed the guest tools directly without reboot. I will a bit more tests later.

**Comment by imtrobin** [comment]
Ok, got a working repro test
1. Fresh Install
2. Do not reboot as it will somehow pull a version from somewhere else. Install Guest tools immediately. 
3. It will popup with host requiring a reboot. Reboot. 
4. It will popup another reboot required. Reboot. It will show Management Agent installed. All working well.

So I wasn't drunk when I installed the first time., The reboot before install guest tool is the one causing the problem. In subsequent installs, I shutdown the VM to do a snapshot.

**Comment by borzel** [comment]
@imtrobin @olivierlambert it's installing thr Citrix drivers from Windows Update. If you install also our drivers windows gehts confused and yoz are in trouble.

How to avoid that?
Sone times ago I reserved the pci device number for XCP-ng for exact that reason. So that VMs with that pci device number instead of the Citrix one is created and prevent the citrix drivers from windows update.

**Comment by borzel** [comment]
@olivierlambert reference: https://github.com/xcp-ng/xcp/issues/76

**Comment by olivierlambert** [comment]
@imtrobin yeah just follow the installer and keep the I/O driver ticked and only reboot when asked, it will work.

**Comment by Gatak** [comment]
Hi. So, haven't followed this for a little while. But if I have a VM migrated from Citrix Xenserver 6, which has the Citrix client tools  installed (Windows 2012R2). Will your installer remove them before proceeding? And how do I prevent the Windows update versions? Or do I need both (mamagement vs drivers?). 

Thank you.

**Comment by synq** [comment]
I can confirm that when installing Windows Server 2019: if you install the XCP-ng Windows PV-Drivers right after installing Windows it works.

If you first update windows totally using Windows update you will get stuck with INACCESSIBLE_BOOT_DEVICE.

**Comment by TristisOris** [comment]
same here. clean install with all updates.
server 2016, 1607 14393.3115

![изображение](https://user-images.githubusercontent.com/1270117/62695683-7bf4d800-b9df-11e9-8eb7-80c9c4176a9a.png)
[Install.log](https://github.com/xcp-ng/xcp/files/3481183/Install.log)

**Comment by armouredking** [comment]
This does not work in Windows 7 v6.1.7601.16385 with 8.2.2.00-rc1 tools (which was the official Microsoft download as of 08/30). It fails to install, rebooting shows errors in Windows EventViewer and the tools report not installed. Attempting to manually install the drivers by pointing Windows to the XCP-ng folder results in the below:

![image](https://user-images.githubusercontent.com/16038428/64068298-187b5600-cbeb-11e9-9640-c027ad285f30.png)

![image](https://user-images.githubusercontent.com/16038428/64068346-afe0a900-cbeb-11e9-93f6-b3c2407d528d.png)

Additionally, unlike the setup.exe file which fails silently, attempting to install the drivers manually gets you the certificate warning about Vates but _STILL FAILS_ even if you click to accept the Vates certificate due to error 52. I was not able to get BCDEDIT disabling to work either with the drivers.

The network still thinks it's Fast Ethernet and behaves as such, not allowing download speeds to get above an abysmal crawl making an update next to impossible unless you want to die of old age first.

Performing a slipstream of updates up to Aug2019 using Integrate7 (v1.7 of the script, v6.1.7601.24511 of Windows) to compile them will still give you an installation fail (yes, even if you reboot it), but at least you can manually introduce the Vates cert after installation and get the drivers to install/load (by manually clicking on them to install / update via Device Manager). Management Agent is still not recognised by the host following the repair, I had to manually install each driver individually to get it working.

**Comment by nagilum99** [comment]
It is like it's ever been and everywhere explaned: The speed shown is just a string - it's not more, not less. Absolutely just bunch of characters displayed. That also means: Nothing anywhere is actively limiting networkspeed - it's just limited by computing power. Also means: Of course the fully emulated NICs are way slower than Xen devices.

**Comment by varnav** [comment]
Is there a fix for system that fails to boot with INACCESSIBLE_BOOT_DEVICE ?

**Comment by holy22** [comment]
The problem with INACCESSIBLE_BOOT_DEVICE is really annoying:(

**Comment by rjt** [comment]
Maybe i missed it in this long thread, so forgive me for using my phone.

 The Windows system account and therefore windows system installer would
not recognize an administrators personal store, so someone tried installing
the certs into the machine level store(s), correct?

Offhand, certlm.msc?

If you want to run as the system user, from an elevated powershell:
whoami
choco install sysinternals
psexec.exe /accepteula /si cmd.exe
whoami

The whoami run inside that elevated cmd.exe should return “nt
authority/system” or similar.

**Comment by borzel** [comment]
@varnav @holy22 @rjt you dont have to deal with certificates, the driver are singed with an valid EV-codesigning cert that is crosssigned from MS.

The reason for INACCESSIBLE_BOOT_DEVICE has to be some other issue. Please follow the discussion at https://xcp-ng.org/forum/topic/2222/error-0xc000000e-on-a-p2v-windows-2016

**Comment by borzel** [comment]
I added a point to my list: "Check if CITRIX-Drivers are present and abort installer to avoid mixing the drivers resulting in an unbootable VM"

This would prevent non-booting VMs like people reported on the forum.

**Comment by borzel** [comment]
added another point: "Allow installation and use without "Window Update" enabled (aka without the additional CITRIX PCI vendor device)"

**Comment by michi53** [comment]
Hi, 
after having trouble, installing the guest-tools to a few vms, it seems, that I have found a workaround, that maybe help someone. 
Before the first time, I saw the desktop of a fresh Win 10 Pro (1909) installation, there came the dialogue, that tells you to restart because of XenServer PV Storage drivers - after installing the XCP-ng-tools (with or without reboot), I get the INACCESSIBLE_BOOT_DEVICE  error.

To get around this, I installed a VM without any network adapter - installed the XCP-ng-tools and then connected to the internet, to get the Windows-Updates. Everything works fine.

**Comment by stormi** [comment]
Actually you probably shouldn't install the XCP-ng tools and at the same time activate Windows Update, since it may update to Citrix drivers that have different IDs and won't work with our agent in the current state of things.

**Comment by Gatak** [comment]
This question has come up before, but isnt it better long-term to use xcp-ng tools so that we can be sure the drivers are supported and compatible? Recent Citrix drivers updates really borked one machine just last week.

**Comment by stormi** [comment]
We're working on it, but the thing is the sources for Citrix drivers are not available. In theory they are better tested, though they don't manage yet to provide a seamless experience, it seems. 

The sources that are available and that we use for XCP-ng drivers are from the Xen project. Citrix drivers are based on them but they also add patches, mostly - they said - to add backwards compatibility.

We are looking for a (paid) dev who is not afraid of Windows so that we could put serious work on that part and not just let the community (that is, @borzel alone on limited free time) handle it. And also be able to debug the cases where things break, be it using our drivers or those from Citrix.

If Citrix driver updates break a machine, there's a chance that ours would too at some point. It would not be honest from me to claim that ours are better at this point. They're reported to work well, but not better.

Right now, you can you either drivers but don't mix them, that's all that is required. And transitioning from one set of tools to the other on an already installed VM requires deep cleaning of the registry and such. One of my colleagues put together a extensive list of things to clean that we will put on the wiki (first), then later ideally automate that directly in our installer.

**Comment by swchok** [comment]
Hi All, 

I did 2 fresh installs of Window Server 2019 Standard (Desktop) Build 17763.

1st install - Chose UEFI as boot, tried to install Windows Management Agent, failed to install. Rebooted with a BSOD, "inaccessible boot device"

2nd install - Chose BIOS, installed and removed windows management agent. Realised in device manager, Xenbus and the other networking drivers will self-install. Updated and reboot, no issues, but windows management agent will still show not installed in XCP-ng center.

**Comment by stormi** [comment]
You need to choose between XCP-ng drivers and Citrix drivers. XenCenter and thus XCP-ng Center enable a feature that automatically installs drivers from Citrix (but not the management agent), that are incompatible with ours and with our management agent.

Two solutions:
* create the VMs from XCP-ng Center and install the management agent from Citrix
* create the VMs from Xen Orchestra and install the drivers+management agent from XCP-ng

**Comment by swchok** [comment]
@stormi 

Samuel, ALAS, I understand it now. I have been using XCP-ng Center. Now I will delete the VMs and startover from Xen Orchestra. Bye bye Citrix.

**Comment by stormi** [comment]
Also make sure you're not using a custom template that has the "Windows Update Tools" option activated. If you're using the default templates, no pb.

**Comment by swchok** [comment]
Thanks @stormi 

And everyone else, stay safe.

**Comment by marxencc** [comment]
If anyone ends up with a production VM stuck in an INACCESSIBLE_BOOT_DEVICE state - i have just fixed my Windows Server 2016 vm by doing the following:

- Get the "Drivers" folder from the XCP Tools installation path (C:\PROGRAM FILES...) - install the tools somewhere else to get it. 
- Create an ISO-Image containing the drivers folder (google for imgburn) and mount that ISO-Image to your VM
- Boot to recovery mode and use the command line and the tool "dism" to inject the drivers (specifically the xenbus and xenvbd drivers) - watch out for the drive letter (D in my case)..
 It should read sth like: 
              dism /image:d:\ /add-driver /driver:e:\Drivers\xenbus\x64\xenbus.inf
              dism /image:d:\ /add-driver /driver:e:\Drivers\xenvbd\x64\xenvbd.inf

cheers, marxen

**Comment by olivierlambert** [comment]
Hi!

That's interesting, would you like to provide those details in our official documentation? (Eg in this section: https://xcp-ng.org/docs/guests.html#windows juste before "Using the Windows guest tools from Citrix").

**Comment by marxencc** [comment]
Hi,

yes, of course! i think it qould be quickest if you just merged it - or shall i provide a more manual-like version? :)

**Comment by olivierlambert** [comment]
In the end of each page, there's a link to create a pull request with your change automatically. This way, you'll know how to modify and next time, you'll be able to do it entirely yourself. So I think it will be nice if you can contribute directly!

**Comment by beshleman** [comment]
@borzel We're facing some issues with the driver installation with secure boot enabled, I was wondering if it wouldn't be too much trouble to ping you about it (any help would be much appreciated).

After a fresh Windows 10 install, `signtool.exe verify /v /kp /path/to/xcp-ng-pv-drivers-installer` fails with:

![no-cert-installed-signtool-verify](https://user-images.githubusercontent.com/6470669/121272367-a2555a00-c87a-11eb-9ce4-fde03c674811.png)

After installing the (now expired) certificate "Certum Trusted Network CA" (thumbprint: 55 43 55 15 fd d2 48 65 75 fd c5 cf 3b ad 00 c9 13 12 3d 03) as a Trusted Root CA in the windows cert manager, and invoking `signtool.exe verify` again, it succeeds:

![screen-after-installing-certum-cert-to-trusted-root-store](https://user-images.githubusercontent.com/6470669/121272423-cdd84480-c87a-11eb-9c5e-2db01e321141.png)

But... after enabling secure boot, verifying again fails, but with a different error message:

![screenshot-signtool-verify-after-sb-enabled](https://user-images.githubusercontent.com/6470669/121272454-e0527e00-c87a-11eb-8066-ec4a2c1df9b8.png)

It claims that `The signing certificate is not valid for the requested purpose`, but yet, the properties of the Vates signing certificate clearly state code signing as one would expect.  This leaves me to wonder, is this because the certificate in the Cross Certificate chain is expired?  Am I configuring the cert store incorrectly?

I also noticed that the Certrum Trusted Network CA that Windows ships with is different than the one used in the successful verification I screenshotted above, likely due to its deprecation in April.

**Comment by csc-techgroup** [comment]
Have published a fix of dead VM here: 

https://xcp-ng.org/forum/topic/2203/windows-server-2019-xcp-ng-windows-pv-drivers-8-2-2-200-rc1-after-install-fail-to-boot/5?loggedin=true
